Support array shuffling with an index array generated with np.random.permutation

agayev opened the pull request:
Currently, it assumes the array that is being indexed is a one-dimensional array of ints and appears to work.  Please see if this makes sense.  If it does, then I will extend it to support multi-dimensional arrays.

ehsantn commented on hpat/distributed_analysis.py:
```diff
@@ -518,6 +518,14 @@ def _analyze_getitem(self, inst, lhs, rhs, array_dists):
+            arr_def = guard(get_definition, self.func_ir, index_var)
+            if (arr_def.op == 'call' and
+                    self._call_table[arr_def.func.name] == ['permutation', 'random', np]):
+                self._meet_array_dists(lhs, rhs.value.name, array_dists)
```
`arr_def` could be None or any node (not necessarily Expr). This `if` needs more checks otherwise it fails.

ehsantn replied:
We need to check `rhs.value` to be multi-dimensional.

ehsantn replied:
call_table is gone now. need to use find_callname.

ehsantn commented on hpat/distributed.py:
```diff
@@ -1099,25 +1124,33 @@ def f(A, val, start, stop, chunk_start, chunk_count):  # pragma: no cover
                 is_multi_dim = True
 
+            arr_def = guard(get_definition, self.func_ir, index_var)
+            if (arr_def.op == 'call' and
```
same issue with `arr_def` check before.

ehsantn commented on hpat/distributed_analysis.py:
```diff
@@ -231,7 +230,7 @@ def _analyze_call(self, lhs, rhs, func_var, args, array_dists):
-        if fdef == ('permutation', 'np.random'):
+        if fdef == ('permutation', 'numpy.random'):
             assert len(args) == 1
             assert self.typemap[args[0].name] == types.int64
```
this needs removed.

ehsantn commented on hpat/distributed.py:
```diff
@@ -676,6 +677,30 @@ def f(lhs, n):
+    def _run_permutation_array_index(self, lhs, rhs, idx):
+        scope, loc = lhs.scope, lhs.loc
+        alloc = mk_alloc(self.typemap, self.calltypes, lhs,
+                         tuple(self._array_sizes[lhs.name]),
```
chunk size of first dimention: self._array_counts[lhs.name][0].

agayev commented:
Fixed the chunk size.  You were wrong about the code path not being exercised.  It was being exercised, except when you execute your shuffle algorithm with one node, you end up sending all of the elements to yourself in sorted order.  The missing thing was shuffle after MPI_Alltoallv.  I added a hack to receive make up a PRNG and shuffle with that, but the proper way is to obtain the seed from numba and use that.

agayev commented:
Dead code eliminator removes np.random.seed() call, which makes it hard to write test for this.  Fixed it in #2873.

ehsantn commented on hpat/_distributed.cpp:
```diff
@@ -470,6 +481,80 @@ void permutation_int(int64_t* output, int n)
 }
 
+std::vector<int> find_dest_ranks(int rank, int num_ranks,
+                                 int64_t *idx, int idx_len)
```
needs doc.

ehsantn commented on hpat/_distributed.cpp:
```diff
@@ -470,6 +481,80 @@ void permutation_int(int64_t* output, int n)
+                  MPI_INT64_T, MPI_COMM_WORLD);
+
+    std::mt19937 rng(0);
+    std::shuffle(ilhs, ilhs + dest_ranks.size(), rng);
```
shuffle? idx should determine the order.

ehsantn commented on hpat/tests/test_basic.py:
```diff
@@ -264,5 +264,16 @@ def test_impl(n):
+            B = np.random.permutation(n)
+            A = A[B]
+            return A.sum()
+
```
Needs a test where two arrays are permuted with the same index array, to make sure they are permuted in the same way.

ehsantn commented on hpat/_distributed.cpp:
```diff
@@ -470,6 +481,80 @@ void permutation_int(int64_t* output, int n)
+
+    MPI_Alltoallv(send_buf.data(), send_counts.data(), send_disps.data(),
+                  MPI_INT64_T, ilhs, recv_counts.data(), recv_disps.data(),
+                  MPI_INT64_T, MPI_COMM_WORLD);
```
Needs support for non-int types.

agayev commented:
The last patch addresses all the issues, except the unit tests.  You can review this, while I work on the unit tests.

ehsantn commented on hpat/_distributed.cpp:
```diff
@@ -470,6 +480,130 @@ void permutation_int(int64_t* output, int n)
+// |num_ranks|, finds the destination ranks of indices of the |rank|.  For
+// example, if |rank| is 1, |num_ranks| is 3, |p_len| is 12, and |p| is the
+// following array [ 9, 8, 6, 4, 11, 7, 2, 3, 5, 0, 1, 10], the function returns
+// [2, 0, 1, 0].
```
Shouldn't this be `[0,2,0,1]`?

agayev replied:
Correct.  Fixed.

ehsantn commented on hpat/_distributed.cpp:
```diff
@@ -470,6 +480,130 @@ void permutation_int(int64_t* output, int n)
+void apply_permutation(unsigned char *v, int64_t elem_size,
+                       std::vector<size_t>& p)
+{
+    std::vector<unsigned char> tmp(elem_size);
```
tmp not used.

agayev replied:
Removed.

ehsantn commented on hpat/_distributed.cpp:
```diff
@@ -470,6 +480,130 @@ void permutation_int(int64_t* output, int n)
+    std::vector<unsigned char> tmp(elem_size);
+    for (size_t i = 0; i < p.size(); ++i) {
+        auto current = i;
+        while (i != p[current]) {
```
This loop needs doc/reference.

agayev replied:
Added.

ehsantn commented on hpat/_distributed.cpp:
```diff
@@ -470,6 +480,130 @@ void permutation_int(int64_t* output, int n)
+    }
+
+    MPI_Alltoallv(send_buf.data(), send_counts.data(), send_disps.data(),
+                  MPI_INT64_T, lhs, recv_counts.data(), recv_disps.data(),
```
Needs MPI data type of size `elem_size`.

agayev replied:
Fixed.

ehsantn commented on hpat/_distributed.cpp:
```diff
@@ -470,6 +480,130 @@ void permutation_int(int64_t* output, int n)
+
+    auto begin = p + hpat_dist_get_start(p_len, num_ranks, rank);
+    auto p1 = arg_sort(begin, dest_ranks.size());
+    auto p2 = arg_sort(p1.data(), dest_ranks.size());
```
Needs explanation.

agayev replied:
Added.

ehsantn commented:
Looks good. Merging.
